### PR TITLE
fix a bug in vw modular

### PR DESCRIPTION
**Pull request by yorkerlin**

### CODE REVIEW

File: examples/undocumented/libshogun/streaming_vwfeatures.cpp [hunk 1]
```diff
@@ -56,7 +59,6 @@ int main()
 
 	SG_UNREF(test_features);
 	SG_UNREF(test_file);
-	SG_UNREF(train_features);
```
Comment (yorkerlin) [inline examples/undocumented/libshogun/streaming_vwfeatures.cpp]: Note that `COnlineSVMSGD` will delete `train_features`.

File: src/shogun/io/streaming/InputParser.h [hunk 2]
```diff
@@ -534,7 +534,7 @@ template <class T> void* CInputParser<T>::main_parse_loop(void* params)
 		pthread_testcancel();
 
 		current_example = examples_ring->get_free_example();
-		current_feature_vector = current_example->fv;
```
Comment (yorkerlin) [inline src/shogun/io/streaming/InputParser.h]: the pointer is NULL, which is a bug.

**Comment by yorkerlin** [comment]
@karlnapf @vigsterkr

**Comment by yorkerlin** [comment]
There is another bug in `VowpalWabbit.cpp`. I will try to fix it.

File: src/shogun/io/streaming/InputParser.h [hunk 3]
```diff
@@ -534,6 +534,7 @@ template <class T> void* CInputParser<T>::main_parse_loop(void* params)
 		pthread_testcancel();
 
 		current_example = examples_ring->get_free_example();
+		current_example->init_fv();
 		current_feature_vector = current_example->fv;
```
Comment (yorkerlin) [inline src/shogun/io/streaming/InputParser.h]: Note: without calling `init_fv()`. `current_example->fv` is a NULL pointer, which is a bug.

File: examples/undocumented/libshogun/CMakeLists.txt [hunk 4]
```diff
@@ -35,7 +35,6 @@ LIST(REMOVE_ITEM EXAMPLES_CPP
 	"streaming_onlinesvmsgd.cpp"
 	"streaming_stringfeatures.cpp"
 	"streaming_vowpalwabbit.cpp"
-	"streaming_vwfeatures.cpp"
```
Comment (karlnapf) [inline examples/undocumented/libshogun/CMakeLists.txt]: why this?

Comment (karlnapf) [inline examples/undocumented/libshogun/CMakeLists.txt]: ah because you fixed it.
Is is valgrinded?

File: src/shogun/io/streaming/ParseBuffer.h [hunk 5]
```diff
@@ -48,6 +48,10 @@ class Example
 	/// Feature vector of type T
 	T* fv;
 	index_t length;
+	void init_fv() 
+	{
```
Comment (karlnapf) [inline src/shogun/io/streaming/ParseBuffer.h]: This should maybe be unit test covered?
Also, is this memory clean?

**Comment by yorkerlin** [comment]
@karlnapf 
yes. 
The following unit test is passed.
```
ctest -R libshogun-streaming_vwfeatures  -D ExperimentalMemCheck 
```

File: examples/undocumented/libshogun/CMakeLists.txt (hunk 4 above)
Comment (yorkerlin) [inline examples/undocumented/libshogun/CMakeLists.txt]: yes

File: src/shogun/io/streaming/ParseBuffer.h [hunk 6]
```diff
@@ -174,6 +174,11 @@ template <class T> class CParseBuffer: public CSGObject
 	 */
 	virtual const char* get_name() const { return "ParseBuffer"; }
 
+	/** Initialize vector if free_vectors_on_destruct is True and the vector is NULL
+	 *
+	 */
+	void init_vector();
```
Comment (yorkerlin) [inline src/shogun/io/streaming/ParseBuffer.h]: This method fixes the bug.

File: src/shogun/io/streaming/ParseBuffer.h [hunk 7]
```diff
@@ -262,7 +277,7 @@ template <class T> CParseBuffer<T>::~CParseBuffer()
 		{
 			SG_DEBUG("%s::~%s(): destroying examples ring vector %d at %p\n",
 					get_name(), get_name(), i, ex_ring[i].fv);
-			SG_FREE(ex_ring[i].fv);
+			delete ex_ring[i].fv;
```
Comment (karlnapf) [inline src/shogun/io/streaming/ParseBuffer.h]: why delete and not SG_FREE?

**Comment by karlnapf** [comment]
Fine to merge from my side

File: src/shogun/io/streaming/ParseBuffer.h [hunk 8]
```diff
@@ -220,6 +225,18 @@ template <class T> class CParseBuffer: public CSGObject
...
 
+
+template <class T> void CParseBuffer<T>::init_vector()
+{
+	if (!free_vectors_on_destruct)
+		return;
+	for (int32_t i=0; i<ring_size; i++)
+	{
+		if(ex_ring[i].fv==NULL)
+			ex_ring[i].fv = new T();
```
Comment (yorkerlin) [inline src/shogun/io/streaming/ParseBuffer.h]: It seems that `SG_CALLOC` does not work. I have to use `new`.

File: examples/undocumented/libshogun/streaming_vwfeatures.cpp [hunk 9]
```diff
@@ -55,12 +58,10 @@ int main()
 		SG_SPRINT("For example %d, predicted label is %f.\n", i, test_labels->get_label(i));
 
 	SG_UNREF(test_features);
-	SG_UNREF(test_file);
-	SG_UNREF(train_features);
-	SG_UNREF(train_file);
```
Comment (yorkerlin) [inline examples/undocumented/libshogun/streaming_vwfeatures.cpp]: @karlnapf 
Note that `CStreamingVwFeatures` will automatically free `train_file`. (in this example)

However, `CStreamingStringFeatures` will `not` automatically free `train_file`. (ref:  shogun/examples/undocumented/libshogun/streaming_stringfeatures.cpp)
We should make them work consistently.

File: src/shogun/io/streaming/ParseBuffer.h (hunk 8 above)
Comment (karlnapf) [inline src/shogun/io/streaming/ParseBuffer.h]: what about malloc then?
Also, details please.
BTW new in shogun is overloaded I think (or it used to be)

File: src/shogun/features/streaming/StreamingStringFeatures.h [hunk 10]
```diff
@@ -257,9 +257,6 @@ template <class T> class CStreamingStringFeatures : public CStreamingFeatures
 	/// If remapping is enabled, this is the target alphabet
 	CAlphabet* alpha_bin;
 
-	/// The StreamingFile object to read from.
-	CStreamingFile* working_file;
```
Comment (yorkerlin) [inline src/shogun/features/streaming/StreamingStringFeatures.h]: this is a bug.

File: examples/undocumented/libshogun/streaming_vwfeatures.cpp (hunk 9 above)
Comment (yorkerlin) [inline examples/undocumented/libshogun/streaming_vwfeatures.cpp]: this issue is fixed due to the bug at src/shogun/features/streaming/StreamingStringFeatures.h

**Comment by yorkerlin** [comment]
The following unit tests are passed.
```
ctest -R libshogun-streaming_vwfeatures  -D ExperimentalMemCheck 
ctest -R libshogun-streaming_stringfeatures  -D ExperimentalMemCheck 
```

File: src/shogun/io/streaming/ParseBuffer.h (hunk 8 above)
Comment (yorkerlin) [inline src/shogun/io/streaming/ParseBuffer.h]: I tried `SG_MALLOC(T, 1);`, it does not work.
Maybe the size/length is wrong. Do you know how can I get the right size given a generic type `T`?

Comment (yorkerlin) [inline src/shogun/io/streaming/ParseBuffer.h]: I tried `SG_MALLOC(T, 1)`, `SG_CALLOC(T, 1)`, `SG_MALLOC(T, sizeof(T) )`, and `SG_CALLOC(T, sizeof(T) )`; and SG_FREE is used. All these attempts are failed.
I have to use `new` and `delete`.

Comment (karlnapf) [inline src/shogun/io/streaming/ParseBuffer.h]: ok then! thanks for checking

**Comment by karlnapf** [comment]
Squash and go ahead and merge.
Please put a meaningful commit message! Say what it does and give a hint why

**Pull request merged**